core/stateless/witness.go

```diff
@@ -18,21 +18,61 @@ package stateless
 
 import (
 	"errors"
+	"fmt"
 	"maps"
 	"slices"
 	"sync"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
 )
 
-// HeaderReader is an interface to pull in headers in place of block hashes for
-// the witness.
+// HeaderReader is an interface to pull in headers in place of block hashes for the witness.
 type HeaderReader interface {
-	// GetHeader retrieves a block header from the database by hash and number,
+	// GetHeader retrieves a block header from the database by hash and number.
 	GetHeader(hash common.Hash, number uint64) *types.Header
 }
 
+// ValidateWitnessPreState validates that the witness pre-state root matches the parent block's state root.
+func ValidateWitnessPreState(witness *Witness, headerReader HeaderReader, peerID string) error {
+	if witness == nil {
+		return fmt.Errorf("witness is nil")
+	}
+
+	// Check if witness has any headers.
+	if len(witness.Headers) == 0 {
+		return fmt.Errorf("witness has no headers")
+	}
+
+	// Get the witness context header (the block this witness is for).
+	contextHeader := witness.Header()
+	if contextHeader == nil {
+		return fmt.Errorf("witness context header is nil")
+	}
+
+	// Get the parent block header from the chain.
+	parentHeader := headerReader.GetHeader(contextHeader.ParentHash, contextHeader.Number.Uint64()-1)
+	if parentHeader == nil {
+		return fmt.Errorf("parent block header not found: parentHash=%x, parentNumber=%d",
+			contextHeader.ParentHash, contextHeader.Number.Uint64()-1)
+	}
```

Comment:
One concern of this approach is that it might prevent a client from bulk downloading witnesses. We might need to move this to the place where the witness is actually being used for importing its block (in `blockchain.go`), since at that point we knew the block's parent must have been known at that point.